### PR TITLE
LAST keyword support 

### PR DESCRIPTION
**Pull request by ea-rus**
## Description

Added LAST keyword. It is created for handling new data from sources and organize pipelines to process new pieces of data

Simple example:
```sql
select * from db.table 
where id > last
```
First run of this query
- executes query `select id from db.table order by id desc limit 1` and will find current max value  for `id`.
- returns empty data

Second run of the query:
- finds current stored value for id
- put it in in query instead of last
- executes query and gets new max value for id from it
User will see only delta of data that appeared in table since previous run of the query.

Variables are stored in the context, current list of contexts:
- job context. different for every job
- view context. different for every view . if view is executed inside of the job it still has view context
- default. 

Because new context variable should be updated this type of query won't work because it doesn't have id on output 
```sql
select name from db.table 
where id > last
```

Other examples:

1. Job to
- get new data from home_rentals
- apply prediction
- insert result to home_rentals_res
```sql
create job j1 (
  insert into my_pg.home_rentals_res
  (
    SELECT
     t.rental_price as real_price, m.rental_price as predicted_price,  t.id
    FROM my_pg.home_rentals as t
    JOIN home_rentals_model as m
    where t.id > last
  )
)
every minute;
```

2. View which returns only new data on every call
```sql
create view v1(
select * from my_pg.branches where id > last
)
```

**Demo**: https://www.loom.com/share/399dec869e0b4438946c5c716926fa7e

***Design doc**: https://docs.google.com/document/d/1hKAGLcibH7apt-QlIO_Rgv2Wi_YiPsEQIEKgfdjlSSQ/edit#heading=h.lvdv8b942awy

Linked PR: https://github.com/mindsdb/mindsdb_sql/pull/313
Fixes: https://github.com/mindsdb/mindsdb_private/issues/234
Fixes: #7104

## Type of change
- [x] ⚡ New feature (non-breaking change which adds functionality)

## Verification Process

To ensure the changes are working as expected:

 - [ ]   Test Location: Specify the URL or path for testing.
 - [ ]   Verification Steps: Outline the steps or queries needed to validate the change. Include any data, configurations, or actions required to reproduce or see the new functionality.

## Additional Media:

- [ ] I have attached a brief loom video or screenshots showcasing the new functionality or change.

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have appropriately commented on my code, especially in complex areas.
- [ ] Necessary documentation updates are either made or tracked in issues.
- [x] Relevant unit and integration tests are updated or added.

### CODE REVIEW

**Review by StpMax — CHANGES_REQUESTED** [review]

File: mindsdb/integrations/handlers/cloud_spanner_handler/README.md [hunk 1]
```diff
@@ -27,6 +27,7 @@ parameters={
     "instance_id":"my-instance",
     "database_id":"example-id",
     "project":"my-project",
+    "dialect": "postgres" // optinonal, default is 'googlesql'
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/integrations/handlers/cloud_spanner_handler/README.md]: fixed

File: mindsdb/api/mysql/mysql_proxy/classes/sql_query.py [hunk 2]
```diff
@@ -570,15 +570,21 @@ def fill_params(node, **kwargs):
...
 
+            if context_callback:
+                context_callback(data, columns_info)
+
+        # is it used?
         # if this is query: execute it
-        if isinstance(data, ASTNode):
-            subquery = SQLQuery(data, session=self.session)
-            return subquery.fetched_data
+        # if isinstance(data, ASTNode):
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/api/mysql/mysql_proxy/classes/sql_query.py]: removed

File: mindsdb/interfaces/query_context/context_controller.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        db.session.commit()
+        return rec
+
+    def __update_context_record(self, context_name, query_str, values):
+        rec = self.__get_context_record(context_name, query_str)
+        rec.values = values
+        db.session.commit()
+
+
+contextController = ContextController()
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/query_context/context_controller.py]: name replaced to query_context_controller

File: mindsdb/interfaces/database/projects.py [hunk 4]
```diff
@@ -114,8 +114,8 @@ def query_view(self, query: ASTNode) -> ASTNode:
             name=view_name,
             project_name=self.name
         )
-        subquery_ast = parse_sql(view_meta['query'], dialect='mindsdb')
-        return subquery_ast
+        view_meta['query_ast'] = parse_sql(view_meta['query'], dialect='mindsdb')
+        return view_meta
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/database/projects.py:118]: No, it was bug, i removed it.

File: mindsdb/api/mysql/mysql_proxy/datahub/datanodes/project_datanode.py [hunk 5]
```diff
@@ -91,16 +92,22 @@ def query(self, query=None, native_query=None, session=None):
         view_query_ast = self.project.query_view(query)
 
         renderer = SqlalchemyRender('mysql')
-        query_str = renderer.get_string(view_query_ast, with_failback=True)
+        view_meta = renderer.get_string(view_query_ast, with_failback=True)
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/api/mysql/mysql_proxy/datahub/datanodes/project_datanode.py]: removed render here, I think we can use AST instead

File: mindsdb/interfaces/query_context/context_controller.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+from mindsdb_sql.parser.ast import Identifier, Select, OrderBy, NullConstant, Constant, BinaryOperation
+
+from mindsdb.interfaces.storage import db
+from mindsdb.utilities.context import context as ctx
+
+from .last_query import LastQuery
+
+
+class ContextController:
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/query_context/context_controller.py]: added docstrings for new classes

File: mindsdb/interfaces/storage/db.py [hunk 7]
```diff
@@ -431,3 +434,16 @@ def as_dict(self) -> Dict:
             'updated_at': self.updated_at,
             'created_at': self.created_at
         }
+
+
+class QueryContext(Base):
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/storage/db.py:439]: done

File: mindsdb/interfaces/query_context/last_query.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+        self.query_in = query
+
+        query = copy.deepcopy(self.query_in)
+
+        last_tables = self.__find_last_columns(query)
+        if last_tables is None:
+            return
+
+        self.query = query
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/query_context/last_query.py:28]: replaced, now are:
- query_orig is original query
- query is modified

File: mindsdb/interfaces/query_context/last_query.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        self.query = None
+
+        # check query type
+        if not isinstance(query, Select):
+            # just skip it
+            return
+
+        self.query_in = query
+
+        query = copy.deepcopy(self.query_in)
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/query_context/last_query.py]: changed: deepcopy is executed after analyzing of query and only if query contents LAST

**Review by StpMax — APPROVED** [review]

**Pull request merged**